stac_geoparquet/arrow/_to_arrow.py

```diff
@@ -47,67 +51,59 @@ def parse_stac_items_to_arrow(
-        # else it would be possible for a STAC item late in the collection (after the
-        # first chunk) to have a different schema and not match the schema inferred for
-        # the first chunk.
-        table = pa.Table.from_batches([_stac_items_to_arrow(items)])
-
-    return _process_arrow_table(table, downcast=downcast)
+        yield _stac_items_to_arrow(items, downcast=downcast)
 
 
-def parse_stac_ndjson_to_arrow(
-    path: Union[str, Path],
+def parse_stac_items_to_arrow(
+    items: Iterable[Dict[str, Any]],
     *,
     chunk_size: int = 8192,
     schema: Optional[pa.Schema] = None,
     downcast: bool = True,
 ) -> pa.Table:
-    # Define outside of if/else to make mypy happy
-    items: List[dict] = []
-
-    # If the schema was not provided, then we need to load all data into memory at once
-    # to perform schema resolution.
-    if schema is None:
-        with open(path) as f:
-            for line in f:
-                items.append(json.loads(line))
-
-        return parse_stac_items_to_arrow(items, chunk_size=chunk_size, schema=schema)
+    batches = parse_stac_items_to_batches(
+        items, chunk_size=chunk_size, schema=schema, downcast=downcast
+    )
+    return pa.Table.from_batches(batches, schema=schema)
 
-    # Otherwise, we can stream over the input, converting each batch of `chunk_size`
-    # into an Arrow RecordBatch at a time. This is much more memory efficient.
-    with open(path) as f:
-        batches: List[pa.RecordBatch] = []
-        for line in f:
-            items.append(json.loads(line))
 
-            if len(items) >= chunk_size:
-                batches.append(_stac_items_to_arrow(items, schema=schema))
-                items = []
+def parse_stac_ndjson_to_batches(
+    path: Union[str, Path],
+    *,
+    chunk_size: int = 8192,
+    schema: Optional[pa.Schema] = None,
+    downcast: bool = True,
+) -> Iterable[pa.RecordBatch]:
+    return parse_stac_items_to_batches(
+        read_json(path), chunk_size=chunk_size, schema=schema, downcast=downcast
+    )
 
-    # Don't forget the last chunk in case the total number of items is not a multiple of
-    # chunk_size.
-    if len(items) > 0:
-        batches.append(_stac_items_to_arrow(items, schema=schema))
 
-    table = pa.Table.from_batches(batches, schema=schema)
-    return _process_arrow_table(table, downcast=downcast)
+def parse_stac_ndjson_to_arrow(
+    path: Union[str, Path],
+    *,
+    chunk_size: int = 8192,
+    schema: Optional[pa.Schema] = None,
+    downcast: bool = True,
+) -> pa.Table:
+    batches = parse_stac_items_to_batches(
+        read_json(path), chunk_size=chunk_size, schema=schema, downcast=downcast
+    )
+    return pa.Table.from_batches(batches, schema=schema)
 
 
-def _process_arrow_table(table: pa.Table, *, downcast: bool = True) -> pa.Table:
+def _process_arrow_table(
+    table: Union[pa.Table, pa.RecordBatch], *, downcast: bool = True
+) -> Union[pa.Table, pa.RecordBatch]:
```

Comment:
I'd rather not have the union here; I'd rather switch this to only work on record batches